internal/resolver/dns/dns_resolver_test.go

```diff
@@ -1490,55 +1498,63 @@ func TestRateLimitedResolve(t *testing.T) {
 		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
 	}
 
-	// Observe the time before unblocking the lookupHost call.  The 100ms rate
-	// limiting timer will begin immediately after that.  This means the next
-	// resolution could happen less than 100ms if we read the time *after*
-	// receiving from tr.ch
-	start := time.Now()
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+
 
 	// Wait for the first resolution request to be done. This happens as part
-	// of the first iteration of the for loop in watcher() because we call
-	// ResolveNow in Build.
-	<-tr.ch
-
-	// Here we start a couple of goroutines. One repeatedly calls ResolveNow()
-	// until asked to stop, and the other waits for two resolution requests to be
-	// made to our testResolver and stops the former. We measure the start and
-	// end times, and expect the duration elapsed to be in the interval
-	// {wantCalls*dnsResRate, wantCalls*dnsResRate}
-	done := make(chan struct{})
-	go func() {
-		for {
-			select {
-			case <-done:
-				return
-			default:
-				r.ResolveNow(resolver.ResolveNowOptions{})
-				time.Sleep(1 * time.Millisecond)
-			}
-		}
-	}()
+	// of the first iteration of the for loop in watcher().
+	select {
+	case <-tr.ch:
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for lookup() call.")
+	}
 
-	gotCalls := 0
-	const wantCalls = 3
-	min, max := wantCalls*dnsResRate, (wantCalls+1)*dnsResRate
-	tMax := time.NewTimer(max)
-	for gotCalls != wantCalls {
-		select {
-		case <-tr.ch:
-			gotCalls++
-		case <-tMax.C:
-			t.Fatalf("Timed out waiting for %v calls after %v; got %v", wantCalls, max, gotCalls)
-		}
+	// Call Resolve Now 100 times, shouldn't continue onto next iteration of watcher, thus shouldn't lookup again.
+	for i := 0; i <= 100; i++ {
+		r.ResolveNow(resolver.ResolveNowOptions{})
+	}
+
+	shouldNotHappenCtx, shouldNotHappenCtxCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer shouldNotHappenCtxCancel()
+
+	select {
+	case <-tr.ch:
+		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
+	case <-shouldNotHappenCtx.Done():
+	}
+
+	timer, err := timerChan.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Error receiving timer from mock NewTimer call: %v", err)
+	}
+	timerPointer := timer.(*time.Timer)
+	timerPointer.Reset(0)
```

Comment:
`time` package's documentation for `Reset()` has the following:
```
// For a Timer created with NewTimer, Reset should be invoked only on
// stopped or expired timers with drained channels.
```

It also states that:
```
// This should not be done concurrent to other receives from the Timer's
// channel.
```

Reply:
I'm wondering what might be the best way out. One option might be as follows:

In `dns_resolver.go`
```
type timer struct {
  t *time.Timer
}

func (t *timer) Stop() {
  t.t.Stop()
}

func (t *timer) Done() <-chan time.Time {
  return t.t.C
}

type timerInterface interface {
  Stop() bool
  Done() <-chan time.Time
}

func newTimer = func(d time.Duration) *time.Timer {
  return &timer{t: time.NewTimer(d)}
}
```

In `dns_resolver_test.go` create a fake timer implementation and override `newTimer()`. Does feel like an overkill, but I'm a little concerned about violating the documentation around `timer.Reset()`.

Reply:
@dfawley : What do you think?

Reply:
This timer.Reset idea was all Doug's which I implemented so defer to him.

Reply:
These are "should"s.  I believe they are there to address potential race issues that might occur under normal timer usage.  In our case, we initially have an infinite timer that will never fire.  I don't think we need to be concerned about anything here.  If you want to be paranoid, you can call `Stop` first, but I really don't think that's worth it.

Reply:
How about we make a helper function, something like
```golang
func waitForNewTimerAndReset(ctx context.Context, ch *testutils.Channel) error {
	t, err := ch.Receive(ctx)
	if err != nil {
		return fmt.Errorf("error receiving timer from mock timer channel: %v", err)
	}
	tp := t.(*time.Timer)
	tp.Reset(0)
	return nil
}
```

and have a comment describing exactly what we are doing.

We already have this pattern multiple times in the existing test and this PR only adds more.

Reply:
It seems this only saves one line, though?

```go
if err := waitForNewTimerAndReset(ctx, myChan); err != nil {
	t.Fatal(...)
}
```
vs
```go
if t, err := ch.Receive(ctx); err != nil {
	t.Fatal(...)
}
t.(*time.Timer).Reset(0)
```

Reply:
I prefer it not as a util function but inline.

Reply:
My reason for having a utility function was more to do with adding a comment or two saying stuff like "alright, we are doing a Reset() with 0, this will make the timer fire immediately, and unblock the resolver waiting on this". But I'm fine either ways.

Reply:
Fair point. I handled explanation in the two tests (well technically one test, and I added onto another) when I moved exponential backoff to this component this way: I put a comment in the function that I set newTimer to as follows: "// Will never fire on its own, allows this test to call timer immediately." This comment is also present in this test, when I overrode newTimerDNSResRate. I agree with your concern about no context for reader to understand what the is happening, but my view is that the comment within the overriden function is good enough if you read it from top to bottom.

Reply:
Since Doug was fine with the level of explanation of my Exponential backoff tests, what I'll do is I'll add a comment to this test explaining the underlying logic, but I don't like the util function idea.

Reply:
Yeah, rereading the test there's no context to what's happening there, you're right. Let me add a comment.

Reply:
Added comment.